### DISCUSSION TITLE
How does the Bevy Resources System work?

### DISCUSSION THREAD

**Discussion by MasonJohnHawver42** [bevyengine/bevy#4690] [Q&A] [answered]
What data structures are used to create the bevy resources system? or here in the source code is this system outlined? I want to recreate this system in C++ for my own game engine. My current implementation uses an array of void pointers and a hash map that maps strings to the index of a void ptr.

**Comment by Geams** [comment, marked as answer]
Search for `insert_resource` in[ bevy::ecs::World](https://github.com/bevyengine/bevy/blob/main/crates/bevy_ecs/src/world/mod.rs).
A resource is a component without an entity attached.
Other projects might also be useful:

- [HECS](https://github.com/Ralith/hecs): Bevy's ECS is based on that
- [Fyrox](https://github.com/FyroxEngine/Fyrox): 3D ECS game engine with a Scene Editor

Good luck with your engine!
